File: backend/src/entropia/application/queries/trash.py
"""Trash read model (Stage 6c, doc 20 §4, §5, §13).

Admin-only cursor projection over ``trash_entries`` — NOT the V18 local
``deletedItems`` array. Every entry point re-applies ``require_trash_admin``
(the route guard is never the only check). Filters (``q``, ``object_type``) and
the stable ``deleted_at DESC, id DESC`` keyset are pushed down to SQL; an
unknown object type or malformed cursor is rejected, never coerced. The detail
query returns the redacted, size-bounded deletion/dependency snapshots only —
no secrets, no raw artifact payloads (doc 20 §12).
"""

from __future__ import annotations

from datetime import datetime
from typing import Any, NamedTuple

from sqlalchemy.ext.asyncio import AsyncSession

from entropia.domain.agent_lab.cursor import clamp_limit
from entropia.domain.identity import Actor
from entropia.domain.identity.policy import require_trash_admin
from entropia.domain.trash.page import (
    DEFAULT_LIST_STATUSES,
    MAX_SEARCH_TEXT,
    TRASH_OBJECT_LOCATIONS,
    TrashEntryStatus,
    decode_trash_cursor,
    encode_trash_cursor,
    normalize_object_type,
)
from entropia.domain.trash.restore import (
    RestoreConflictKind,
    conflict_summary,
    resolution_options,
)
from entropia.infrastructure.postgres.models import TrashEntry
from entropia.infrastructure.postgres.repositories import entities as entity_repo
from entropia.infrastructure.postgres.repositories import trash as trash_repo
from entropia.shared.errors import CursorInvalidError, TrashEntryNotFoundError

_RESULT_ENTITY_TYPE = "backtest_result"
_ARTIFACT_ENTITY_TYPE = "hypothesis_artifact"
_MANUAL_ENTITY_TYPE = "manual_document"

_HEAD_POINTER_MOVED = RestoreConflictKind.HEAD_POINTER_MOVED
_TARGET_MISSING = RestoreConflictKind.TARGET_MISSING

# Statuses where restore is closed by the deletion lifecycle itself — there is no
# resolution to offer, only the error code the command would raise (doc 20 §4).
_BLOCKED_REASON_BY_STATUS: dict[TrashEntryStatus, str] = {
    TrashEntryStatus.PURGE_PENDING: "PURGE_IN_PROGRESS",
    TrashEntryStatus.PURGED: "OBJECT_ALREADY_PURGED",
    TrashEntryStatus.RESTORED: "ENTITY_NOT_SOFT_DELETED",
}

# An entry is restore-eligible when its object is recoverable-soft-deleted:
# purge_failed returned the root to soft_deleted; purge_pending disables Restore
# (doc 20 §4 interaction matrix).
_RESTORE_ELIGIBLE_STATUSES = (TrashEntryStatus.SOFT_DELETED, TrashEntryStatus.PURGE_FAILED)

_PURGE_STATUS_BY_ENTRY_STATUS: dict[TrashEntryStatus, str | None] = {
    TrashEntryStatus.SOFT_DELETED: None,
    TrashEntryStatus.RESTORED: None,
    TrashEntryStatus.PURGE_PENDING: "pending",
    TrashEntryStatus.PURGE_FAILED: "failed",
    TrashEntryStatus.PURGED: "completed",
}


def _row(entry: TrashEntry) -> dict[str, Any]:
    return {
        "trash_entry_id": entry.id,
        "entity_id": entry.entity_id,
        "object_type": entry.entity_type,
        "display_name": entry.display_name or entry.entity_id,
        "original_location": entry.original_location,
        "original_owner": entry.owner_at_deletion,
        "deleted_by": entry.deleted_by,
        "deleted_at": entry.deleted_at.isoformat() if entry.deleted_at else None,
        "delete_reason": entry.reason,
        "status": str(entry.status),
        "purge_status": _PURGE_STATUS_BY_ENTRY_STATUS.get(entry.status),
        "purge_job_id": entry.purge_job_id,
        "restore_eligible": entry.status in _RESTORE_ELIGIBLE_STATUSES,
        "row_version": entry.row_version,
        "correlation_id": entry.correlation_id,
    }


async def list_trash_entries(
    session: AsyncSession,
    actor: Actor,
    *,
    q: str | None = None,
    object_type: str | None = None,
    cursor: str | None = None,
    limit: int | None = None,
) -> dict[str, Any]:
    """Admin-only, filtered, newest-deletion-first cursor page (doc 20 §4, §13)."""
    require_trash_admin(actor)
    page_limit = clamp_limit(limit)
    otype = normalize_object_type(object_type)
    needle = (q or "").strip()[:MAX_SEARCH_TEXT] or None

    last: tuple[datetime, str] | None = None
    if cursor is not None:
        deleted_at_iso, entry_id = decode_trash_cursor(cursor)
        try:
            last = (datetime.fromisoformat(deleted_at_iso), entry_id)
        except ValueError as exc:
            raise CursorInvalidError() from exc

    rows = await trash_repo.list_entries(
        session,
        statuses=DEFAULT_LIST_STATUSES,
        object_type=otype,
        q=needle,
        last=last,
        limit=page_limit + 1,
    )
    has_more = len(rows) > page_limit
    page = list(rows[:page_limit])
    next_cursor: str | None = None
    if has_more and page:
        tail = page[-1]
        next_cursor = encode_trash_cursor(
            deleted_at_iso=tail.deleted_at.isoformat(), entry_id=tail.id
        )

    recoverable_total = await trash_repo.count_recoverable(session)
    return {
        "data": [_row(r) for r in page],
        "meta": {
            "cursor": next_cursor,
            "has_more": has_more,
            "limit": page_limit,
            "recoverable_total": recoverable_total,
            "object_types": sorted(TRASH_OBJECT_LOCATIONS),
        },
    }


async def _target_deletion_state(session: AsyncSession, entry: TrashEntry) -> str | None:
    """Best-effort CURRENT deletion state of the target, for refresh semantics
    (doc 20 §7 Refresh status). ``None`` when the target row is unresolvable."""
    if entry.entity_type == _RESULT_ENTITY_TYPE:
        from entropia.infrastructure.postgres.repositories import backtest as bt_repo

        result = await bt_repo.get_result(session, entry.entity_id)
        return result.deletion_state if result is not None else None
    if entry.entity_type == _ARTIFACT_ENTITY_TYPE:
        from entropia.infrastructure.postgres.repositories import agent_lab as al_repo

        artifact = await al_repo.get_hypothesis(session, entry.entity_id)
        return str(artifact.deletion_state) if artifact is not None else None
    root = await entity_repo.get_root(session, entry.entity_id)
    return str(root.deletion_state) if root is not None else None


async def get_trash_entry_detail(
    session: AsyncSession, actor: Actor, *, trash_entry_id: str
) -> dict[str, Any]:
    """Admin-only snapshot reader (doc 20 §3.2, §7 Open Snapshot).

    Returns the immutable redacted deletion/dependency snapshots recorded at
    delete time plus the live purge/restore control state. Read-only: no root
    reactivation, no snapshot edit.
    """
    require_trash_admin(actor)
    entry = await trash_repo.get_entry(session, trash_entry_id)
    if entry is None:
        raise TrashEntryNotFoundError()

    detail = _row(entry)
    tombstone = await trash_repo.get_tombstone(session, entry.entity_id)
    detail.update(
        {
            "deletion_snapshot": entry.deletion_snapshot or {},
            "dependency_snapshot": entry.dependency_snapshot or {},
            "purge_error": entry.purge_error,
            "purge_requested_by": entry.purge_requested_by,
            "restored_at": entry.restored_at.isoformat() if entry.restored_at else None,
            "restored_by": entry.restored_by,
            "current_deletion_state": await _target_deletion_state(session, entry),
            "tombstone": (
                {
                    "purged_at": tombstone.purged_at.isoformat() if tombstone.purged_at else None,
                    "purged_by": tombstone.purged_by,
                }
                if tombstone is not None
                else None
            ),
        }
    )
    return detail


# --------------------------------------------------------------------------- #
# Restore preflight (O-17, doc 20 §5 "Restore conflict choice", §8.2)          #
# --------------------------------------------------------------------------- #


class _TargetProbe(NamedTuple):
    """Read-only view of what the restore command would find.

    ``adjudicates_head`` is what separates "this type pins a head pointer" from
    "its head happens to be NULL" — collapsing the two would silently skip the
    historical-integrity check for a root whose pointer really is unset.
    """

    resolved: bool
    adjudicates_head: bool = False
    current_head: Any = None


async def _probe_restore_target(session: AsyncSession, entry: TrashEntry) -> _TargetProbe:
    """Mirror of the command's per-type target lookup — WITHOUT any mutation.

    Only ``EntityRegistry`` roots and manual documents adjudicate a head pointer
    on restore; a Result / Lab artifact carries its deletion flag on its own row
    and has no snapshot head to disagree with (commands/deletion.py).
    """
    if entry.entity_type == _RESULT_ENTITY_TYPE:
        from entropia.infrastructure.postgres.repositories import backtest as bt_repo

        result = await bt_repo.get_result(session, entry.entity_id)
        return _TargetProbe(resolved=result is not None)
    if entry.entity_type == _ARTIFACT_ENTITY_TYPE:
        from entropia.infrastructure.postgres.repositories import agent_lab as al_repo

        artifact = await al_repo.get_hypothesis(session, entry.entity_id)
        return _TargetProbe(resolved=artifact is not None)
    if entry.entity_type == _MANUAL_ENTITY_TYPE:
        from entropia.infrastructure.postgres.repositories import manual as manual_repo

        document = await manual_repo.get_document(session, entry.entity_id)
        if document is None:
            return _TargetProbe(resolved=False)
        return _TargetProbe(True, adjudicates_head=True, current_head=document.current_revision_id)
    root = await entity_repo.get_root(session, entry.entity_id)
    if root is None:
        return _TargetProbe(resolved=False)
    return _TargetProbe(True, adjudicates_head=True, current_head=root.current_revision_id)


def _conflict_report(
    kind: RestoreConflictKind, *, expected_head: Any = None, current_head: Any = None
) -> dict[str, Any]:
    """The typed conflict + its option set, from the ONE domain catalog."""
    return {
        "conflict_kind": str(kind),
        "conflict_summary": conflict_summary(
            kind, expected_head=expected_head, current_head=current_head
        ),
        "resolutions": resolution_options(kind),
    }


async def get_restore_preflight(
    session: AsyncSession, actor: Actor, *, trash_entry_id: str
) -> dict[str, Any]:
    """Admin-only READ answering "would Restore succeed, and if not, what may I
    choose?" (O-17; doc 20 §5 "Typed ``resolution`` enum from preflight", §8.2).

    Read-only by construction: no lock, no mutation, no partial projection. The
    Admin drives the doc 20 §6 "Restore needs attention" panel from this, then
    resubmits the restore with the chosen typed ``resolution`` and the SAME
    ``expected_head_revision_id`` this response echoes.

    ``outcome`` is one of:

    * ``allow``    — preflight passes; Restore may be submitted as-is.
    * ``conflict`` — a typed conflict, with the resolutions its domain adapter
      supports. An EMPTY ``resolutions`` list is a real answer, not an omission:
      nothing can unblock it and the object stays in Trash (doc 20 §8.2).
    * ``blocked``  — the deletion lifecycle forbids restore outright (purge in
      flight, already purged, already restored). ``blocked_reason`` carries the
      error code the command would raise; there is nothing to choose.

    Preflight is advisory, never authorization: the command re-runs every check
    inside its own transaction, so a race between this read and the submit is
    caught there (and by OCC), not trusted from here.
    """
    require_trash_admin(actor)
    entry = await trash_repo.get_entry(session, trash_entry_id)
    if entry is None:
        raise TrashEntryNotFoundError()

    preflight: dict[str, Any] = {
        "trash_entry_id": entry.id,
        "entity_id": entry.entity_id,
        "object_type": entry.entity_type,
        "display_name": entry.display_name or entry.entity_id,
        # Echoed so the retry carries the SAME expected deletion version the
        # Admin reviewed; a moved entry then fails OCC instead of restoring blind.
        "expected_head_revision_id": entry.row_version,
        "outcome": "allow",
        "blocked_reason": None,
        "conflict_kind": None,
        "conflict_summary": None,
        "resolutions": [],
    }

    blocked_reason = _BLOCKED_REASON_BY_STATUS.get(entry.status)
    if blocked_reason is not None:
        preflight.update({"outcome": "blocked", "blocked_reason": blocked_reason})
        return preflight

    probe = await _probe_restore_target(session, entry)
    if not probe.resolved:
        preflight.update({"outcome": "conflict", **_conflict_report(_TARGET_MISSING)})
        return preflight

    expected_head = (entry.dependency_snapshot or {}).get("current_revision_id")
    if probe.adjudicates_head and expected_head is not None and expected_head != probe.current_head:
        preflight.update(
            {
                "outcome": "conflict",
                **_conflict_report(
                    _HEAD_POINTER_MOVED,
                    expected_head=expected_head,
                    current_head=probe.current_head,
                ),
            }
        )
    return preflight


__all__ = ["get_restore_preflight", "get_trash_entry_detail", "list_trash_entries"]

File: frontend/src/lib/trash.ts
// Admin Trash data access (Stage 6c doc 20 §7): the recoverable soft-deleted
// entry index, the immutable per-entry snapshot detail, and the OCC-guarded
// restore. Every surface is Admin-only at the ROUTE and again in the service
// (require_trash_admin) — a hidden nav item is never authorization (doc 20 §2);
// a non-Admin sees the server 403 envelope verbatim.
//
// Trash has no dedicated SSE event: a restore changes an entity's lifecycle
// (resource.changed → full refresh) and emits an audit event
// (audit.event.created → ["audit"]). Read keys live under ["trash"]; the
// restore mutation invalidates ["trash"] + ["audit"] the same way
// lib/adminPanel.ts useAssignRole does.

import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";

import { ApiError, api, apiRequest } from "./apiClient";

// ---------------------------------------------------------------------------
// Wire types (mirror backend application/queries/trash.py `_row`/detail and
// application/commands/deletion.py restore return verbatim)
// ---------------------------------------------------------------------------

export interface TrashEntry {
  trash_entry_id: string;
  entity_id: string;
  object_type: string;
  display_name: string;
  original_location: string | null;
  original_owner: string | null;
  deleted_by: string | null;
  deleted_at: string | null;
  delete_reason: string | null;
  status: string;
  purge_status: string | null;
  purge_job_id: string | null;
  restore_eligible: boolean;
  row_version: number;
  correlation_id: string | null;
}

export interface TrashEntriesPage {
  data: TrashEntry[];
  meta: {
    cursor: string | null;
    has_more: boolean;
    limit: number;
    recoverable_total: number;
    object_types: string[];
  };
}

export interface TrashTombstone {
  purged_at: string | null;
  purged_by: string | null;
}

export interface TrashEntryDetail extends TrashEntry {
  deletion_snapshot: Record<string, unknown>;
  dependency_snapshot: Record<string, unknown>;
  purge_error: string | null;
  purge_requested_by: string | null;
  restored_at: string | null;
  restored_by: string | null;
  current_deletion_state: string | null;
  tombstone: TrashTombstone | null;
}

export interface RestoreResult {
  trash_entry_id: string;
  entity_id: string;
  entity_type: string;
  display_name: string;
  status: string;
  deletion_state: string;
  current_revision_id: string | number | null;
  // O-17: the typed resolution that actually unlocked this restore, or null when
  // it restored cleanly and nothing needed resolving.
  applied_resolution: string | null;
  row_version: number;
  correlation_id: string | null;
}

// ---------------------------------------------------------------------------
// Restore preflight + typed conflict resolutions (O-17, doc 20 §5, §6, §8.2)
// ---------------------------------------------------------------------------

// One option the DOMAIN adapter offers. The client never invents, renames or
// filters these — an empty option set means nothing can unblock the conflict
// and the object stays in Trash (doc 20 §8.2).
export interface RestoreResolutionOption {
  conflict_kind: string;
  resolution: string;
  label: string;
  description: string;
}

export interface RestorePreflight {
  trash_entry_id: string;
  entity_id: string;
  object_type: string;
  display_name: string;
  expected_head_revision_id: number;
  outcome: "allow" | "conflict" | "blocked";
  blocked_reason: string | null;
  conflict_kind: string | null;
  conflict_summary: string | null;
  resolutions: RestoreResolutionOption[];
}

export const RESTORE_CONFLICT_CODE = "RESTORE_CONFLICT";

// The 409 envelope carries the same catalog the preflight read returns: message
// is the {conflict_summary} doc 20 §6 interpolates, details is one row per typed
// resolution. Parsed defensively — a row without a resolution token is dropped
// rather than rendered as a nameless choice.
export function restoreConflictOptions(error: unknown): RestoreResolutionOption[] {
  if (!(error instanceof ApiError) || error.code !== RESTORE_CONFLICT_CODE) return [];
  return error.details.flatMap((row) => {
    const resolution = typeof row.resolution === "string" ? row.resolution : null;
    if (resolution === null) return [];
    return [
      {
        conflict_kind: typeof row.conflict_kind === "string" ? row.conflict_kind : "",
        resolution,
        label: typeof row.label === "string" ? row.label : resolution,
        description: typeof row.description === "string" ? row.description : "",
      },
    ];
  });
}

// Purge request 202 return (mirrors commands/deletion.py request_purge dict
// verbatim): the target moved to purge_pending and a durable maintenance job
// is enqueued (doc 20 §8.3). The purge itself runs in the worker — the request
// only accepts the job.
export interface PurgeResult {
  purge_job_id: string;
  trash_entry_id: string;
  entity_id: string;
  entity_type: string;
  deletion_state: string;
  purge_status: string;
  row_version: number;
  correlation_id: string | null;
}

// ---------------------------------------------------------------------------
// Filters — the object_type option list is hydrated from the server response
// (`meta.object_types`), never a hard-coded client list.
// ---------------------------------------------------------------------------

export interface TrashFilters {
  q: string | null;
  object_type: string | null;
}

export const DEFAULT_TRASH_FILTERS: TrashFilters = { q: null, object_type: null };

// Badge tone for the live purge/restore control state (presentation only — the
// wire status/purge_status stay plain strings).
export function purgeStatusTone(purgeStatus: string | null): "warn" | "down" | "neutral" {
  if (purgeStatus === "pending") return "warn";
  if (purgeStatus === "failed" || purgeStatus === "purged") return "down";
  return "neutral";
}

// ---------------------------------------------------------------------------
// Query hooks — all under the ["trash"] prefix (swept by resource.changed)
// ---------------------------------------------------------------------------

export function useTrashEntries(filters: TrashFilters, cursor: string | null) {
  return useQuery({
    queryKey: ["trash", "entries", filters.q, filters.object_type, cursor],
    queryFn: () => {
      const params = new URLSearchParams();
      if (filters.q) params.set("q", filters.q);
      if (filters.object_type) params.set("object_type", filters.object_type);
      if (cursor !== null) params.set("cursor", cursor);
      const qs = params.toString();
      return api.get<TrashEntriesPage>(`/trash-entries${qs ? `?${qs}` : ""}`);
    },
    // Keep the current table mounted while a page flip is in flight.
    placeholderData: (previous) => previous,
  });
}

export function useTrashEntry(trashEntryId: string | null) {
  return useQuery({
    queryKey: ["trash", "entry", trashEntryId],
    queryFn: () =>
      api.get<TrashEntryDetail>(`/trash-entries/${encodeURIComponent(trashEntryId ?? "")}`),
    enabled: trashEntryId !== null,
  });
}

// Restore preflight (O-17, doc 20 §5, §8.2) — a READ, so no OCC token and no
// Idempotency-Key. Disabled until an entry is actually being adjudicated; the
// Admin uses it to re-read the canonical conflict after the record changed
// ("Admin refreshes canonical detail", doc 20 §8.2). Advisory only: the restore
// command re-runs every check itself, so this is never treated as permission.
export function useRestorePreflight(trashEntryId: string | null) {
  return useQuery({
    queryKey: ["trash", "restore-preflight", trashEntryId],
    queryFn: () =>
      api.get<RestorePreflight>(
        `/trash-entries/${encodeURIComponent(trashEntryId ?? "")}/restore-preflight`,
      ),
    enabled: trashEntryId !== null,
    // Always re-read on demand: a cached preflight would defeat its purpose.
    staleTime: 0,
    gcTime: 0,
  });
}

// ---------------------------------------------------------------------------
// Mutation — restore (POST /trash-entries/{id}/restore)
// ---------------------------------------------------------------------------

// OCC: expected_head_revision_id carries the entry's row_version so a stale tab
// gets the 409 envelope verbatim instead of restoring against a moved target.
// A fresh Idempotency-Key per attempt keeps a retry after a rejection a new
// decision, not a replay (the body token wins over If-Match, doc 20 §14).
//
// O-17: `resolution` is the typed choice the Admin picked from the conflict's
// option set — sent ONLY when one was chosen, and only ever a token the server
// itself offered. The retry reuses the SAME expected_head_revision_id, so a
// record that moved in the meantime fails OCC instead of restoring blind
// (doc 20 §8.2). An unknown token is rejected 422 server-side; the client never
// substitutes, defaults or drops it.
export function useRestoreEntry() {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: (input: {
      trash_entry_id: string;
      expected_head_revision_id: number;
      resolution?: string;
    }) =>
      apiRequest<RestoreResult>(
        `/trash-entries/${encodeURIComponent(input.trash_entry_id)}/restore`,
        {
          method: "POST",
          headers: { "Idempotency-Key": crypto.randomUUID() },
          body: {
            expected_head_revision_id: input.expected_head_revision_id,
            ...(input.resolution !== undefined ? { resolution: input.resolution } : {}),
          },
        },
      ),
    onSuccess: () => {
      // The root lifecycle changed AND the command emitted an audit event.
      void queryClient.invalidateQueries({ queryKey: ["trash"] });
      void queryClient.invalidateQueries({ queryKey: ["audit"] });
    },
  });
}

// ---------------------------------------------------------------------------
// Mutation — request permanent delete / purge (POST /trash-entries/{id}/purge)
// ---------------------------------------------------------------------------

// Permanent Delete is a two-phase 202 (doc 20 §8.3): the request only moves the
// target to purge_pending and enqueues a durable maintenance job — the worker
// re-checks eligibility and tombstones (or returns it to soft_deleted on
// failure). The command requires BOTH a second confirmation and a re-auth
// proof:
//   - confirmation_phrase must equal the object's display identity
//     (display_name || entity_id) or the server rejects with
//     PURGE_CONFIRMATION_INVALID — never started.
//   - reauth_proof must be a non-empty token from a completed Admin
//     re-authentication step (V1 contract: presence only; full MFA is out of
//     scope, doc 20 §0) or the server rejects with REAUTH_REQUIRED.
// OCC: expected_head_revision_id carries the entry's row_version (body token
// wins over If-Match, doc 20 §14) so a stale tab gets STALE_REVISION verbatim
// instead of purging a moved target. A fresh Idempotency-Key per attempt keeps
// a retry after a rejection a new decision, not a replay (a duplicate submit
// with the SAME key returns the same job, doc 20 §9.3).
export function useRequestPurge() {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: (input: {
      trash_entry_id: string;
      confirmation_phrase: string;
      reauth_proof: string;
      expected_head_revision_id: number;
    }) =>
      apiRequest<PurgeResult>(
        `/trash-entries/${encodeURIComponent(input.trash_entry_id)}/purge`,
        {
          method: "POST",
          headers: { "Idempotency-Key": crypto.randomUUID() },
          body: {
            confirmation_phrase: input.confirmation_phrase,
            reauth_proof: input.reauth_proof,
            expected_head_revision_id: input.expected_head_revision_id,
          },
        },
      ),
    onSuccess: () => {
      // The target moved to purge_pending AND the command emitted an audit event.
      void queryClient.invalidateQueries({ queryKey: ["trash"] });
      void queryClient.invalidateQueries({ queryKey: ["audit"] });
    },
  });
}
